src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyDictionary.cs

```diff
@@ -272,15 +271,15 @@ void ICollection<TKey>.Clear()
 
             public bool Contains(TKey item)
             {
-                return _collection.Contains(item);
+                return _dictionary.ContainsKey(item);
```

Comment:
Yes, effectively it favors paying the price of populating these collections every time rather than reporting inconsistent data.

Reply:
For context: https://github.com/dotnet/runtime/issues/93829

Reply:
But it'll now be inconsistent with ConcurrentDictionary.  If you get the keys collection from ConcurrentDictionary, you'll be getting a snapshot, but if you get it from a wrapping ReadOnlyDictionary, you'll be getting a non-snapshot.  It's definitely a breaking change.  If we're really considering taking such a breaking change, we should consider going all the way and changing CD rather than changing ROD, such that CD.Keys/Values have non-snapshot behavior.

Reply:
> If you get the keys collection from ConcurrentDictionary, you'll be getting a snapshot, but if you get it from a wrapping ReadOnlyDictionary, you'll be getting a non-snapshot.

Is that an issue though? ReadOnlyDictionary is wrapping dictionaries, not their key/value collections. I think it's a bigger problem that ReadOnlyDictionary can be inconsistent w.r.t. its own key/value collections.

Reply:
The getter of `ReadOnlyDictionary.Keys` returning a snapshot is not a problem. The only thing we need to fix is that now it can return an outdated snapshot.

Reply:
With these changes working with `ReadOnlyDictionary.Keys` or `.Values` or in other words working with the `KeyCollection` or `ValueCollection` can degrade performance if populating a new snapshot of `ConcurrentDictionary.Keys`/`.Values` now.
For example if using `Count` of `KeyCollection`  and after that the Enumerator. There are two calls to `_dictionary.Keys` in that case resulting in two populated snapshots.

But as you  already said:

> [...] paying the price of populating these collections every time rather than reporting inconsistent data.

I do not have a better solution right now, so maybe this is the best you can do about it.

Reply:
> Is that an issue though? 

Why wouldn't it be? You're changing from a cached snapshot to a new snapshot on every access.  From a performance perspective, every access to ReadOnlyDictionary.Keys will now be copying the full collection rather than being ammortized allocation-free.  And from a behavior perspective, it means that code like:
```C#
ICollection<TKey> keys = GetReadOnlyDictionary().Keys;
TKey[] array = new TKey[keys.Count];
keys.CopyTo(array, 0);
```
might now blow up in situations it didn't use to, e.g. because the collection might be growing concurrently, but the cached snapshot prevents concurrency-related problems.
 
We might say we think the new behavior is better.  But it's absolutely a breaking change and it should not be made non-chalantly.

Reply:
> might now blow up in situations it didn't use to, e.g. because the collection might be growing concurrently, but the cached snapshot prevents concurrency-related problems.

Wouldn't the same concern apply if we changed ConcurrentDictionary to use live collections rather than snapshots?

Reply:
> Wouldn't the same concern apply if we changed ConcurrentDictionary to use live collections rather than snapshots?

Yes, which is why I said if we're really considering making a breaking change like this, we should consider instead making it in CD.

Reply:
> You're changing from a cached snapshot to a new snapshot on every access. From a performance perspective, every access to ReadOnlyDictionary.Keys will now be copying the full collection rather than being ammortized allocation-free.

That's assuming the underlying dictionary uses snapshot semantics. One alternative approach might be to avoid caching the ROD collections altogether, which would make it more faithful to the underlying type:

```diff
-public KeyCollection Keys => _keys ??= new KeyCollection(m_dictionary.Keys);
+public KeyCollection Keys => new KeyCollection(m_dictionary.Keys);
```

However that would end up penalizing performance in collections that don't use snapshotting.

I'm growing increasingly convinced that the current behavior isn't great when used in conjunction with ConcurrentDictionary. It's not just that it reports stale data, the ROD also has the potential of rooting entries that have long been wiped from the underlying CD.

Reply:
> I'm growing increasingly convinced that the current behavior isn't great when used in conjunction with ConcurrentDictionary.

I'm not disagreeing.  I'm saying this change to address those concerns has its own risks.

Reply:
Somewhat of an ad-hoc solution, but this could work as well:
```C#
private readonly bool _cacheCollections;

public ReadOnlyDictionar(IDictionary<TKey, TValue> dictionary)
{
    _cacheCollections = dictionary.GetType() == typeof(ConcurrentDictionary<TKey, TValue>);
}

public KeyCollection Keys => _cacheCollections ? _keys ??= new(m_dictionary.Keys) :  new(m_dictionary.Keys);
```

Reply:
> Yes, which is why I said if we're really considering making a breaking change like this, we should consider instead making it in CD. Either way, it's a non-trivial impact.

I would expect that a change on the CD collections directly would have substantially more impact, since the example you provided works fine when applied to CD directly:
```C#
ICollection<TKey> keys = concurrentDictionary.Keys;
TKey[] array = new TKey[keys.Count];
keys.CopyTo(array, 0);
```
But is fundamentally broken when applied to a ROD wrapper per #93829, making it less likely for user code to depend on it.

Reply:
Yes, but if your concern is consistency within a type, as you've expressed for ROD, the snapshot behavior of Keys on CD has all the same issues (and is the root cause of the concerns with ROD).

Reply:
It's not exactly the same though, because CD will return a fresh snapshot every time you get it but ROD will cache the first snapshot for its entire lifetime. I actually don't mind the snapshotting semantics of CD, in many cases it might actually be desirable (although I can see how it could trip users up).